### ISSUE TITLE
Support certificate validation

### ISSUE THREAD

**Issue by rmhrisk** [node-saml/xml-crypto#88]
I have a pure Javascript library that supports robust X509 certificate chain validation, it works well in both node and the browser.

You can find the current version here: https://pkijs.org

It is based on WebCrypto, to work on node it uses another library I had put together here : https://github.com/PeculiarVentures/node-webcrypto-ossl

It seems, in general, you have decided to model afer the SignedXML class in .NET (I worked on this while at MS so I think it was a good choice :p)

If so, consider adding a CheckSignature(X509Certificate, Boolean) where you use pkijs to do certificate validation also. You would probably have to also take a certificate store on the method since there is no accessible OS trust store but it could easily be produced with: https://github.com/PeculiarVentures/tl-create

If you are interested in doing this I have a newer version of PKIjs that should be ready for release soon that passes all of the sane tests in http://csrc.nist.gov/groups/ST/crypto_apps_infra/documents/PKITS.pdf that I will release. The forge chain builder won't work in many cases as it rudimentary, PKIjs should work with pretty much all certs.

**Comment by yaronn** [comment]
Yeah, I was a Microsoft mvp for a while on WCF and was very impressed with the signature API, great work!

Same deal as with the other suggestion - hope to get to it but given I have recently started a new job may take a while.

**Comment by rmhrisk** [comment]
I can totally relate :p  this is another 2-3 days of work with some testing I think, I will put it on the list, but again dont wait for us :)

**Comment by bjrmatos** [comment]
i'm curious and i need to ask something..

what is the difference between [PKIJS](https://pkijs.org/) and [node-forge](https://github.com/digitalbazaar/forge)?

**Comment by rmhrisk** [comment]
Forge set out to implement TLS in Javascript, it did this before there was WebCrypto so it imported a bunch of javascript implementations of crypto for those operations (including some insecure algorithm choices even then). For more on Javascript cryptography check out this article: https://tonyarcieri.com/whats-wrong-with-webcrypto 

Its primary goal was to implement TLS so the support for anything outside of TLS was done to a minimal level to enable delivering on that goal. A great example is that of certificate validation, it implements a super rudimentary validation engine that will not work a large majority of the time.

It also had to make due with platform limitations that existed in the browser at the time, for this reason it also took dependencies on things like flash which is thankfully finally dying.

Over the years, other oddball things not related to TLS were added but again only to a minimal level to scratch a specific itch.

PKIjs set out with a different goal, to make it possible to build standards compliant, secure, PKI aware applications in Javascript within the browser utilizing WebCrypto for native crypto implementations. In principle, if PKIjs existed when Forge was created, Forge would have used PKIjs and focused on TLS instead of the PKI elements like certificate processing.

To give you a small idea of the scope differential, the yet to be released version of PKIjs has well over 300 tests for the certificate validation engine, my rough guess is that Forge might pass 20% of those if they were to ran against it.

There are lots of other examples like this also, for example, the message format used for signing PDFs is called CMS, it supports a bunch of options, PKIjs supports them all. It is, for example possible to produce a standards compliant LTV (https://en.wikipedia.org/wiki/PAdES) signed PDF (which is done usinng CMS messages) with PKIjs and that is not possible with Forge.

In short, while Forge is very impressive, it is really more a series of science experiments while PKIjs is aspiring to be a SDK to enable people to build rich PKI-aware applications in both Node and the Browser.

**Comment by bjrmatos** [comment]
thnks for you detailed explanation!

one more question:

> if PKIjs existed when Forge was created, Forge would have used PKIjs and focused on TLS instead of the PKI elements like certificate processing.

do you have plans to ask node-forge's mantainers to rely on PKIjs? i think everybody will win if that happened

**Comment by rmhrisk** [comment]
I would be happy to support them rewriting but I strongly suspect they would have no interest in doing that.

Based on my experience those who use forge do not do so for TLS, they do so for those other bits they had to implement to get a working TLS implementation.

I am also unable to identify a case where a pure JS TLS implementation provides much production value outside some fridge cases weirdos like myself think might be interesting.

Given the above and the lack of activity in their repository I strongly suspect they wouldn't do the work.

Your library on the other hand has a ton of interesting production possibilities :)

**Comment by bjrmatos** [comment]
> Based on my experience those who use forge do not do so for TLS, they do so for those other bits they had to implement to get a working TLS implementation.

you're right i have used it to parse X509 certificates hahaha

again thnks for your toughts.

**Comment by jykae** [comment]
Any status update on this one, is someone able to work on this?

**Comment by rmhrisk** [comment]
We have built on this excellent library and added basic X.509 support : https://www.npmjs.com/package/xadesjs

**Comment by cjbarth** [comment]
@rmhrisk , it seems you no longer have interest in adding to this library to make it work in the browser, so I'll close this. However, we'd still very much welcome a PR with a test suite to accomplish what you've described here. Please consider submitting one.

**Comment by cjbarth** [comment]
See #87
